### PR TITLE
doc on Formations

### PR DESCRIPTION
**Pull request by heinezen**
This PR aims to summarize all results from #705 which @Piruzzolo and me have discovered. I think nobody wants to read through the lengthy discussion to get something done.

- [x] Principles used in AoE 2 
- [x] Formation structure
  - [x] Land Units
  - [x] Ships
- [x] Subformation structure
- [x] Formation Types
- [x] Limitations in AoE2
- [x] Pseudocode/Algorithm proposal

### CODE REVIEW

**Comment by heinezen** [comment]
Ugh finally done... This turned out to one hell of a document.

File: doc/reverse_engineering/formations.md [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+        default: subformation[5].add(u)
+        break();
+    }
+}
+
+subformation[1].order_units(formation_type)
+subformation[2].order_units(formation_type)
+subformation[3].order_units(formation_type)
+subformation[4].order_units(formation_type)
+subformation[5].order_units(formation_type)
```
Comment (TheJJ) [inline doc/reverse_engineering/formations.md]: heh, that is the interesting part :) how does the ordering actually work? and how is it combined with pathfinding? might be something for another pull-request though.

File: doc/reverse_engineering/formations.md [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+Algorithm:
+
+for unit u in player_selection do
+{
+    switch(u.GROUP_ID)
+    {
+        case 912:
+        case 947: subformation[1].add(u)
+        break();
+        case 906: subformation[2].add(u)
```
Comment (TheJJ) [inline doc/reverse_engineering/formations.md]: can you pls do:
``` cpp
for unit u in player_selection do {
    switch (u.GROUP_ID) {
    case 912:
    case 947: subformation[1].add(u);
        break;
    case 906: subformation[2].add(u);
    ....
```

this is more conformant with our code style then.

File: doc/reverse_engineering/formations.md [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+```
+...............
+.1.1.1.1.1.1.1.
+.1.2.2.2.2.2.1.
+.1.2.3.3.3.2.1.
+.1.2.3.4.3.2.1.
+.1.2.3.3.3.2.1.
+.1.2.2.2.2.2.1.
+.1.1.1.1.1.1.1.
+.........
```
Comment (TheJJ) [inline doc/reverse_engineering/formations.md]: missing `......`?

File: doc/reverse_engineering/formations.md (hunk 1 above)
Comment (heinezen) [inline doc/reverse_engineering/formations.md]: shouldn't be that hard to come up with and algorithm. I'll have a look and write it all down..

**Comment by heinezen** [comment]
I've added pseudocode for the `add()` and `order_units()` part of the subformation system but stuck to the behavior already described in the doc. Somebody else has to figure out the pathfinding ;-) The ordering is actually really easy because it basically just iterates through the unit stacks. It's astonishing how simple the implementation seems to be when looked at in detail.

**Comment by TheJJ** [comment]
Ready to merge?

**Comment by heinezen** [comment]
@TheJJ yup :)

**Pull request merged**